### ISSUE TITLE
[4.0] Shadows don't render properly with orthogonal perspective

### ISSUE THREAD

**Issue by megalobyte** [godotengine/godot#58332]
### Godot version

499eec13a3

### System information

Arch, Vukan, X11, NVIDIA GeForce GTX 1070/PCIe/SSE2

### Issue description

Shadows don't appear to be rendered correctly when using an orthogonal perspective. 

Orthogonal:
![Screenshot from 2022-02-19 19-05-44](https://user-images.githubusercontent.com/80655527/154823344-165daca9-b68d-479b-a499-2e6f67a7b327.png)

Perspective: 
![Screenshot from 2022-02-19 19-06-02](https://user-images.githubusercontent.com/80655527/154823345-3ffecb55-7fb9-4a71-aa46-317d1cc0f2c0.png)

Scene uses a directional light. 

### Steps to reproduce

Any scene with a directional light and an orthogonal camera perspective should recreate the issue, but I've attached a project for convenience. 

### Minimal reproduction project

[Uploading minimal_project.zip…]()

**Comment by Calinou** [comment]
Judging by the screenshot, the shadows in orthogonal mode look correct to me. You should probably reduce Shadow Max Distance in the DirectionalLight3D node to increase their effective resolution – the default value is tuned for fairly large open worlds, not small-scale games.

Also, the minimal reproduction project didn't upload (likely because you clicked away from the text field while it was still uploading). Try editing your post and upload it again.

**Comment by megalobyte** [comment]
You are right, except it was the camera's far property that I had to update. Is it too early to update the documentation? I know a lot of things are in flux with 4.0

**Comment by Calinou** [comment]
> You are right, except it was the camera's far property that I had to update.

The Far property defaults to 4000 in `master` to support large open worlds out of the box

The issue here is that DirectionalLight3D does not use the Shadow Max Distance property when the camera is in orthogonal mode, and uses the camera Far value as the maximum distance instead. Can you check whether this happens in 3.4.2 too?

**Comment by megalobyte** [comment]
In 3.4.2, I set the far value to 4000 and this is what I got: 
![Screenshot from 2022-02-19 21-43-26](https://user-images.githubusercontent.com/80655527/154826323-de0c8c82-bb4a-40d1-9980-6f1b431a7992.png)

I should mention that the Max Shadow distance was the same value for 3.4.2 and 4.0

**Comment by Calinou** [comment]
> In 3.4.2, I set the far value to 4000 and this is what I got: ![Screenshot from 2022-02-19 21-43-26](https://user-images.githubusercontent.com/80655527/154826323-de0c8c82-bb4a-40d1-9980-6f1b431a7992.png)
> 
> I should mention that the Max Shadow distance was the same value for 3.4.2 and 4.0

That roughly looks like the shadow resolution you'd get in `master`. You can compare it more easily by setting the directional shadow filter quality to **Hard** in the Project Settings.

**Comment by megalobyte** [comment]
About the same resolution, but doesn't have the same blend effect as 3.4.2
![Screenshot from 2022-02-20 12-30-58](https://user-images.githubusercontent.com/80655527/154856006-47392a47-17af-4c9e-a6d5-3a12214b6ec9.png)

I'm guessing that's probably just a difference with how the rendering engine works?

**Comment by Calinou** [comment]
> I'm guessing that's probably just a difference with how the rendering engine works?

3.4 uses PCF5 shadow filtering by default. To compare its shadow resolution with `master`, you have to set the shadow filter mode to **Disabled** in the Project Settings (in 3.4.2).

**Comment by megalobyte** [comment]
You are right again :)

Thank you for the help! I've learned a lot

**Comment by Calinou** [comment]
I can confirm this on `master` f4b0c7a1e and `3.x` 3ba980379 (both GLES3 and GLES2).

**Testing project (`master`):** 
**Testing project (`3.x`):** [test_shadow_orthogonal_camera_3.x.zip](https://github.com/godotengine/godot/files/8532903/test_shadow_orthogonal_camera_3.x.zip)

## `master`

*The DirectionalLight3D's Shadow Max Distance is set to 100 on all screenshots.*

### Editor Camera Far = 4000

![image](https://user-images.githubusercontent.com/180032/164512951-38c5bc23-b46b-4a24-a451-ce368f17efea.png)

### Editor Camera Far = 500

![image](https://user-images.githubusercontent.com/180032/164512990-a0c8f7c4-8cd7-4eec-93a3-71bb8b93adb4.png)

### Editor Camera Far = 100

![image](https://user-images.githubusercontent.com/180032/164513018-c270e425-0fde-4572-bfbf-a431cd904902.png)


## `3.x`

*The DirectionalLight's Shadow Max Distance is set to 100 on all screenshots.*

### Editor Camera Far = 4000

![image](https://user-images.githubusercontent.com/180032/164512174-a4e8ab03-bed3-4254-8796-39d06cf20c86.png)

### Editor Camera Far = 500

![image](https://user-images.githubusercontent.com/180032/164512275-f7742239-4658-424d-9aaf-035f31bc2c9f.png)

### Editor Camera Far = 100

![image](https://user-images.githubusercontent.com/180032/164512326-165fb332-e614-4a9c-80a8-cd3330956ada.png)

**Comment by clayjohn** [comment]
It looks like shadow_max_distance has been ignored for a long time 
(well over 5 years https://github.com/godotengine/godot/commit/eedb39091aaa1dc0b8f204844bb1eb270b2349f6), so there must be something else causing the issue

**Comment by Calinou** [comment]
This appears to be done on purpose here: https://github.com/godotengine/godot/blob/251fb83d536a96110790bbf85b3fa50e5a8e2b7c/servers/rendering/renderer_scene_cull.cpp#L2084

If you remove the check, shadows will disappear unless you greatly increase Shadow Max Distance in the DirectionalLight3D properties (it must be at least 3,200 or so if the editor camera's Far property is 4,000).

**Comment by golddotasksquestions** [comment]
Many people are reporting Shadow Mesh LOD not working correctly when they try the "Your first 3D game" official tutorial:

https://www.reddit.com/r/godot/comments/170wqon/weird_shadows_in_godot_3d_tutorial/

There seems to be a relation to orthogonal camera setting

**Comment by Calinou** [comment]
@golddotasksquestions This is a different issue, which is tracked in https://github.com/godotengine/godot/issues/78422.

**Comment by RobertBColton** [comment]
Could we add a setting for `Close Fit` or `Stable Fit` for this like Unity has?

Currently facing this issue with an ortho camera that has isometric/dimetric pivot. I disable the znear by setting it to a large negative value, as is typical, because it streamlines switching the camera between perspective and ortho modes.

In Unity they give the option to change the shadow fit to address this problem.
<img width="433" height="178" alt="Unity Shadow Fit" src="https://github.com/user-attachments/assets/d46553cf-c743-4596-bffe-4748fc7769c4" />
https://thinkinginsideadifferentbox.wordpress.com/2020/09/27/orthographic_camera_tips_for_unity3d/

>It’s especially bad for orthographic cameras because the cascade selection is based on some **dodgy heuristic** involving **near** and far **clipping** planes and our **disabling the near plane** on our Orthographic Camera totally **breaks that heuristic**.

**Comment by Calinou** [comment]
> Could we add a setting for Close Fit or Stable Fit for this like Unity has?

This is ultimately a different thing. Close Fit (optimized depth range) and Stable Fit (stable depth range) is about whether you want to favor shadow texel density (and therefore detail) or stability (avoiding pixel crawling when the camera moves or rotates). This was available as a DirectionalLight3D property in 3.x but was [removed in 4.0](https://github.com/godotengine/godot/pull/45023) to make the code simpler, and also because optimized depth range was seldom used in the first place (the pixel crawling was a dealbreaker for most projects). 3.x used the stable depth range by default, and 4.0 always uses stable depth range.

You wouldn't want shadows to also become unstable and suffer from pixel crawling in an orthogonal camera, so a different solution is needed here.

**Comment by RobertBColton** [comment]
I see, the whole problem with ortho projections is that changing the size can make it appear like it's somewhere it's not.

And that messes with a lot more than just shadows.

>In fact, given any specific view on your screen, there are literally an infinite number of possible camera orientations along a vector which will produce the exact same Camera View result. This can fool you into thinking you have a sensible position for your camera when, in fact, your camera could be a million miles away. Or it could be at 0,0,0.
